web/src/emoji_frequency.ts

```diff
@@ -0,0 +1,174 @@
+import assert from "minimalistic-assert";
+
+import * as all_messages_data from "./all_messages_data.ts";
+import * as emoji_picker from "./emoji_picker.ts";
+import * as message_store from "./message_store.ts";
+import * as reactions from "./reactions.ts";
+import {current_user} from "./state_data.ts";
+import * as typeahead from "./typeahead.ts";
+
+export type ReactionData = {
+    score: number;
+    emoji_code: string;
+    message_ids: Set<number>;
+    current_user_reacted_message_ids: Set<number>;
+};
+
+const MAX_FREQUENTLY_USED_EMOJIS = 12;
+const CURRENT_USER_REACTION_WEIGHT = 5;
+export const reaction_data = new Map<string, ReactionData>();
+
+export function update_frequently_used_emojis_list(): void {
+    const frequently_used_emojis = [...reaction_data.values()].toSorted(
+        (a, b) => b.score - a.score,
+    );
+
+    const top_frequently_used_emoji_codes = [];
+    let popular_emojis = [...typeahead.popular_emojis];
+    for (const emoji of frequently_used_emojis) {
+        if (
+            top_frequently_used_emoji_codes.length + popular_emojis.length ===
+            MAX_FREQUENTLY_USED_EMOJIS
+        ) {
+            break;
```

Comment:
This feels very messy with mutating `popular_emojis` partway through. I feel like there should be a cleaner way to implement the union logic we want here.